tutorials/arvo/parsing.md

```diff
@@ -0,0 +1,638 @@
++++
+title = "Parsing tutorial"
+weight = 12
+template = "doc.html"
++++
+
+# Parsing Tutorial
+
+This document serves as an introduction to parsing text with Hoon. No prior
+knowledge of parsing is required, and we will explain the basic structure of how
+parsing works in a purely functional language such as Hoon before moving on to
+how it is implemented in Hoon.
+
+## What is parsing? {#what-is-parsing}
+
+A program which takes a raw sequence of characters as an input and produces a data
+structure as an output is known as a _parser_. The data structure produced
+depends on the use case, but often it may be represented as a tree and the
+output is thought of as a structural representation of the input. Parsers are ubiquitous in
+computing, commonly used for to perform tasks such as reading files, compiling
+source code, or understanding commands input in a command line interface.
+
+Parsing a string is rarely done all at once. Instead, it is usually done
+character-by-character, and the return contains the data structure representing
+what has been parsed thus far as well as the remainder of the string to be
+parsed. They also need to be able to fail in case the input is improperly
+formed. We will see each of these standard practices implemented in Hoon below.
+
+## Functional parsers
+
+How parsers are built varies substantially depending on what sort of programming
+language it is written in. As Hoon is a functional programming language, we will
+be focused on understanding _functional parsers_, also known as _combinator
+parsers_.
+
+Complex functional parsers are built piece by piece from simpler parsers
+that are plugged into one another in various ways to perform the desired task.
+
+The basic building blocks, or primitives, are parsers that read only a
+single character. There are frequently a few types of possible input characters,
+such as letters, numbers, and symbols. For example, `parse(integer, "1")` calls
+the parsing routine on the string `"1"` and looks for an integer, and so it
+returns the integer `1`. However, taking into account what was said above about
+parsers returning the unparsed portion of the string as well, we should
+represent this return as a tuple. So we should expect something like this:
+```
+> parse(integer, "1")
+(1, "")
+> parse(integer, "123")
+(1, "23")
+```
+What if we wish to parse the rest of the string? We would need to apply the
+`parse(integer, -)` function again:
+```
+> parse(integer, parse(integer, "123"))
+(12, "3")
+> parse(integer, parse(integer, parse(integer, "123")))
+(123, "")
```

Comment:
I was confused by what you meant here, since `parse(integer, - )` *is* the same signature to me - integer here being an argument that was already input, and `-` being a blank spot for the string to go in. But I could see how you could see "integer" as meaning the string of the integer to be parsed, so there is potential confusion. The use of `-` to represent a placeholder for an argument is common enough in mathematics literature, but I'm not sure if its common in programming. In any case I'm just going to rewrite it as `parse` by itself, which sidesteps the issue I think